src/main/java/org/cbioportal/web/columnar/ClinicalDataBinner.java

```diff
@@ -0,0 +1,121 @@
+import org.cbioportal.web.parameter.*;
+import org.cbioportal.web.util.DataBinner;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@Component
+public class ClinicalDataBinner {
+    private final StudyViewColumnarService studyViewColumnarService;
+    private final DataBinner dataBinner;
+    
+    @Autowired
+    public ClinicalDataBinner(
+        StudyViewColumnarService studyViewColumnarService,
+        DataBinner dataBinner
+    ) {
+        this.studyViewColumnarService = studyViewColumnarService;
+        this.dataBinner = dataBinner;
+    }
+
+    public List<ClinicalDataBin> fetchClinicalDataBinCounts(
+        DataBinMethod dataBinMethod,
+        ClinicalDataBinCountFilter dataBinCountFilter,
+        boolean shouldRemoveSelfFromFilter
+    ) {
+        List<ClinicalDataBinFilter> attributes = dataBinCountFilter.getAttributes();
+        StudyViewFilter studyViewFilter = dataBinCountFilter.getStudyViewFilter();
+
+        if (shouldRemoveSelfFromFilter) {
+            studyViewFilter = NewClinicalDataBinUtil.removeSelfFromFilter(dataBinCountFilter);
+        }
+
+        List<String> attributeIds = attributes.stream().map(ClinicalDataBinFilter::getAttributeId).collect(Collectors.toList());
+
+        // a new StudyView filter to partially filter by study and sample ids only
+        StudyViewFilter partialFilter = new StudyViewFilter();
+        partialFilter.setStudyIds(studyViewFilter.getStudyIds());
+        partialFilter.setSampleIdentifiers(studyViewFilter.getSampleIdentifiers());
+
+        // filter only by study id and sample identifiers, ignore rest
+        List<Sample> unfilteredSamples = studyViewColumnarService.getFilteredSamples(partialFilter);
+        List<Sample> filteredSamples = studyViewColumnarService.getFilteredSamples(studyViewFilter);
+
+        // TODO make sure these keys don't need to be distinct
+        List<String> unfilteredUniqueSampleKeys = unfilteredSamples.stream().map(Sample::getUniqueSampleKey).collect(Collectors.toList());
+        List<String> filteredUniqueSampleKeys = filteredSamples.stream().map(Sample::getUniqueSampleKey).collect(Collectors.toList());
+        List<String> unfilteredUniquePatientKeys = unfilteredSamples.stream().map(Sample::getUniquePatientKey).collect(Collectors.toList());
+        List<String> filteredUniquePatientKeys = filteredSamples.stream().map(Sample::getUniquePatientKey).collect(Collectors.toList());
+
+        // TODO make sure we don't need a distinction between sample vs patient attribute ids here
+        //  ideally we shouldn't because we have patient clinical data separated from sample clinical data in clickhouse
+        List<ClinicalData> unfilteredClinicalDataForSamples = studyViewColumnarService.getSampleClinicalData(partialFilter, attributeIds);
```

Comment:
also, we should comment here why we do this.  i think it's just because we want to show the filtered bins, right?  Just like the deselected pie charts?  lets use this opportunity to better document the business logic